File: search/bfs/WaterJugs.hs
-- WaterJugs.hs
-- Copyright (C) 2014 Liu Xinyu (liuxinyu95@gmail.com)
--
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

module WaterJugs where

import Debug.Trace

-- General solution to 2 jugs:

-- Problem:
-- Given 2 jugs of volumn a and b (a < b), bring up g volumns of water from the river.

-- Constraint:
-- g must be value of n * gcd(a, b), written as gcd(a, b) | g
-- otherwise, it's unsolvable.
-- a and b are not neccessarily relatively prime, for gcd(a, b) = 1, it's possible to
-- bring up any quantity of g.

-- Memo of extended Euclid algorithm:
-- The extended Euclide algorithm does not only give GCD of a and b, but also give one of
-- the linear combination so that

--     a x + b y = gcd(a, b)

-- Define this function as:

--     (d, x, y) = ext-gcd(a, b)

-- Without loss of generality, assume a < b. there exists quotation q and remainder r that

--    b = a q + r

-- Since d = gcd(a, b), it can divide both a and b, thus d can divide r as well.
-- Because r is less than a, we can scale down the problem to find the GCD of a and r:

--    (d, x', y') = ext-gcd(r, a)

-- Where d = x' r + y' a according to the definition of extended Euclid algorithm.

-- Transform b = a q + r to r = b - a q, substitude r in above equation yields:

--    d = x' (b - a q) + y' a   this can be further transform to:
--    d = (y' - x' q) a + x' b

-- This is linear combination of a and b, so that we have:

--    x = y' - x' (b / a)
--    y = x'

-- Beside this recursive case, the edge case happens when a = 0:
--  gcd(a, b) = b = 0 a + 1 b

-- Summarize the above gives the extended Euclide algorithm:
extGcd 0 b = (b, 0, 1)
extGcd a b = let (d, x', y') = extGcd (b `mod` a) a in (d, y' - x' * (b `div` a), x')

-- Method 1, Extended Euclid algorithm.

-- let g = a x + b y.
-- If a < 0, repeatedly increase x by b and decrese y by a till a is greater than 0.
-- Repeat x times:
--    Fill the jug of volumn a;
--    Pour all the water in jug a into jug b. Whenever jug b becomes full, empty it out.
--solve :: Integer -> Integer -> Integer -> [(Integer, Integer)]
solve a b g | g `mod` d /= 0 = [] -- no solution
            | otherwise = solve' (x * g `div` d)
    where
      (d, x, y) = extGcd a b
      solve' x | x < 0 = solve' (x + b)
               | otherwise = pour x [(0, 0)]
      pour 0 ps = reverse ((0, g):ps)
      pour x ps@((a', b'):_) | a' == 0 = pour (x - 1) ((a, b'):ps) -- fill a
                             | b' == b = pour x ((a', 0):ps) -- empty b
                             | otherwise = pour x ((max 0 (a' + b' - b), min (a' + b') b):ps)

-- Method 2, BFS (brute-force)
solve' a b g = bfs [[(0, 0)]] where
  bfs [] = []
  bfs (c:cs) | fst (head c) == g || snd (head c) == g = reverse c
             | otherwise = bfs (cs ++ map (:c) (expand c))
  expand ((x, y):ps) = filter (`notElem` ps) $ map (\f -> f x y)
                           [fillA, fillB, pourA, pourB, emptyA, emptyB]
  fillA _ y = (a, y)
  fillB x _ = (x, b)
  emptyA _ y = (0, y)
  emptyB x _ = (x, 0)
  pourA x y = (max 0 (x + y - b), min (x + y) b)
  pourB x y = (min (x + y) a, max 0 (x + y - a))

-- Some test data:
-- a = 3, b = 5, g = 4
-- a = 4, b = 9, g = 6

File: search/bfs/waterjugs.c
/*
 * waterjugs.c
 * Copyright (C) 2014 Liu Xinyu (liuxinyu95@gmail.com)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <stdio.h>
#include <stdlib.h>

int min(int a, int b) { return a <= b ? a : b; }

int max(int a, int b) { return b < a ? a : b; }

struct Step {
    int p, q;
    struct Step* parent;
};

struct Step* make_step(int p, int q, struct Step* parent) {
    struct Step* s = (struct Step*) malloc(sizeof(struct Step));
    s->p = p;
    s->q = q;
    s->parent = parent;
    return s;
}

struct Step *steps[1000], **head, **tail = steps;

void push(struct Step* s) { *tail++ = s; }

struct Step* pop() { return *head++; }

int empty() { return head == tail; }

void reset() {
    struct Step **p;
    for (p = steps; p != tail; ++p)
        free(*p);
    head = tail = steps;
}

int eq(struct Step* a, struct Step* b) {
    return a->p == b->p && a->q == b->q;
}

int visited(struct Step* s) {
    struct Step **p;
    for (p = steps; p != tail; ++p)
        if (eq(*p, s)) return 1;
    return 0;
}

void expand(struct Step* s, int a, int b, struct Step** cs) {
    int p = s->p, q = s->q;
    cs[0] = make_step(a, q, s); /*fill A*/
    cs[1] = make_step(p, b, s); /*fill B*/
    cs[2] = make_step(0, q, s); /*empty A*/
    cs[3] = make_step(p, 0, s); /*empty B*/
    cs[4] = make_step(max(0, p + q - b), min(p + q, b), s); /*pour A into B*/
    cs[5] = make_step(min(p + q, a), max(0, p + q - a), s); /*pour B into A*/
}

struct Step* solve(int a, int b, int g) {
    int i;
    struct Step *cur, *cs[6];
    reset();
    push(make_step(0, 0, NULL));
    while (!empty()) {
        cur = pop();
        if (cur->p == g || cur->q == g)
            return cur;
        else {
            expand(cur, a, b, cs);
            for (i = 0; i < 6; ++i)
                if(!eq(cur, cs[i]) && !visited(cs[i]))
                    push(cs[i]);
        }
    }
    return NULL;
}

/**print the steps in reverse order, returns how many steps there are. */
int print(struct Step* s) {
    int n = -1;
    if (s) {
        n = 1 + print(s->parent);
        printf("%d, %d\n", s->p, s->q);
    }
    return n;
}

void test(int a, int b, int g) {
    printf("solve a=%d, b=%d, g=%d:\n", a, b, g);
    printf("total %d steps\n", print(solve(a, b, g)));
    reset();
}

int main(int argc, char *argv[]) {
    test(3, 5, 4);
    test(4, 9, 6);
    return 0;
}
